src/comm.cpp

```diff
@@ -138,6 +138,122 @@ void lbann_comm::split_trainers(
+  num_process_primary_grid = world_size/2; 
+  }
+
+  // Check if parameters are valid
+  if (num_process_primary_grid +  num_process_secondary_grid > world_size) {
+    LBANN_ERROR(
+      "Not enough processes to split the trainer; procs_per_trainer: ",
+      world_size,
+      " is smaller than sum of num processes for primary grid: ",
+      num_process_primary_grid,
+      " and secondary grid: ",
+      num_process_secondary_grid);
+  }
+  if (num_process_primary_grid == 0) {
+    LBANN_ERROR("Procs for primary grid in a trainer cannot be zero.");
+  }
+
+  if(num_process_secondary_grid == 0 and (num_process_primary_grid == world_size)){
+    return ;
+  }
+
+  if(num_process_secondary_grid==0){
+    num_process_secondary_grid = world_size - num_process_primary_grid;
+  }
+
+  int rank_in_split_comm;
+  if(m_rank_in_trainer < num_process_primary_grid){
+    // color = 0;
+    rank_in_split_comm = m_rank_in_trainer % num_process_primary_grid;
+    m_grid_type = PRIMARY_GRID;
+    m_rank_in_trainer = rank_in_split_comm;
+    m_procs_per_trainer = num_process_primary_grid;
+  }
+  else{
+    // color = 1;
+    rank_in_split_comm = (m_rank_in_trainer - num_process_primary_grid) % num_process_secondary_grid;
+    m_grid_type = SECONDARY_GRID;
+    m_rank_in_trainer = rank_in_split_comm;
+    m_procs_per_trainer = num_process_secondary_grid;
+  }
+
+  // Update ranks in primary and secondary grids
+  for (int rank = 0; rank < num_process_primary_grid; ++rank){
+    m_primary_grid_ranks.push_back(rank);
+  }
+  for (int rank = num_process_primary_grid; 
+        rank < num_process_primary_grid + num_process_secondary_grid; 
+        ++rank){
+    m_secondary_grid_ranks.push_back(rank);
+  }
+
+  std::cout<<"Primary Grid:";
+  for (auto it = m_primary_grid_ranks.begin(); it != m_primary_grid_ranks.end(); it++)
+        std::cout << *it << " ";
+  std::cout<<"\n";
+
+  std::cout<<"Secondary Grid:";
+  for (auto it = m_secondary_grid_ranks.begin(); it != m_secondary_grid_ranks.end(); it++)
+        std::cout << *it << " ";
+  std::cout<<"\n";
+
+  //Create Groups to form communicators
+  El::mpi::Group trainer_group, primary_grid_group, secondary_grid_group;
+  El::mpi::CommGroup( m_trainer_comm, trainer_group );
+  El::mpi::Incl( trainer_group, m_primary_grid_ranks.size(), m_primary_grid_ranks.data(), primary_grid_group);
+  El::mpi::Incl( trainer_group, m_secondary_grid_ranks.size(), m_secondary_grid_ranks.data(), secondary_grid_group);
+
+  //Create communicators (one each for primary and secondary grid)
+  El::mpi::Create(m_trainer_comm, primary_grid_group, m_primary_grid_comm);
+  El::mpi::Create(m_trainer_comm, secondary_grid_group, m_secondary_grid_comm);
+
+
+  El::mpi::Dup(m_trainer_comm, m_combined_grid_comm);
+  if(m_create_two_models){
+    if(m_grid_type==PRIMARY_GRID){
+      El::mpi::Dup(m_primary_grid_comm, m_trainer_comm);
+    }
+    else{
+      El::mpi::Dup(m_secondary_grid_comm, m_trainer_comm);
+    }
```

Comment:
I feel uneasy that the trainer comm no longer covers the entire trainer, but only the current subgrid. I understand that this minimizes the impact on the rest of the code. We should document `m_trainer_comm` and mention that the name is misleading.

Reply:
Sure, I will do it.